test: [DI-20585] - Add ACLP Cypress Test Coverage for Linode Dashboard Widgets

agorthi-akamai opened the pull request:
## Description 📝
Added aclp e2e test cases 

## Changes  🔄
List any change relevant to the reviewer.
- ...
- ...

## How to test 🧪
```
yarn cy:run -s cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts
``````

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [ ] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

venkymano-akamai commented on packages/manager/cypress/support/util/cloudpulse.ts:
```diff
@@ -0,0 +1,357 @@
+ * @param {string} region - The name of the region to select.
+ */
+export const selectRegion = (region: string) => {
+  // ui.regionSelect.find().click();
```
These two steps are not working even for existing tests like create-linode and clone-linode
```
ui.regionSelect.find().click();
ui.autocompletePopper.findByTitle(region);
```
Since this is a common thing, Can we please check this?

venkymano-akamai commented on packages/manager/cypress/support/util/cloudpulse.ts:
```diff
@@ -0,0 +1,357 @@
+ */
+export const selectAndVerifyServiceName = (service: string) => {
+  const resourceInput = ui.autocomplete.findByTitleCustom('Select Resources');
+  resourceInput.findByTitle('Open').click();
```
Aftert this (we opened the autocompletepopper now), the below step doesn't work

`ui.autocompletePopper.findByTitle(service);`

jdamore-linode replied:
@venkymano-akamai Happy to take a closer look once this test is passing, but can't offer much advice for now since I can't reproduce.

In the meantime, here's an example in our OBJ access keys test where we have a multi select and select multiple options:

https://github.com/linode/manager/blob/develop/packages/manager/cypress/e2e/core/objectStorage/access-keys.smoke.spec.ts#L225-L231

jdamore-linode commented:
@agorthi-akamai Taking this out of draft status so that the tests can run. I'll take a look soon! Thanks for the PR!

jdamore-linode commented:
@agorthi-akamai Thanks for the PR! I'm seeing these fail to build in CI -- checking it out locally, it looks like there's a stray import in `cypress/support/util/cloudpulse.ts` that either needs to be installed if it's needed, or removed if it's not.

Aside from that, I'm unable to run any of the tests locally because they don't seem to be using any mocks. These tests run against our Production API so Cypress tests for upcoming features need to use mock API data.

agorthi-akamai commented:
> @agorthi-akamai Thanks for the PR! I'm seeing these fail to build in CI -- checking it out locally, it looks like there's a stray import in `cypress/support/util/cloudpulse.ts` that either needs to be installed if it's needed, or removed if it's not.
> 
> Aside from that, I'm unable to run any of the tests locally because they don't seem to be using any mocks. These tests run against our Production API so Cypress tests for upcoming features need to use mock API data.

I've addressed to your code review comments; kindly review

jdamore-linode commented:
@abailly-akamai Heads up that I took this out of draft -- I need this to not be in draft so I can run the E2Es.

jdamore-linode commented:
@abailly-akamai No worries! It makes sense for this to be a draft, just doesn't play nice with our pipeline 😭

venkymano-akamai commented on packages/manager/src/components/Autocomplete/Autocomplete.tsx:
```diff
@@ -159,7 +159,7 @@ export const Autocomplete = <
       PopperComponent={CustomPopper}
       clearOnBlur={clearOnBlur}
-      data-qa-autocomplete
+      data-qa-autocomplete={label}
```
FYI, if we add this label, we will uniquely and easily identify entire autocomplete in a page where we have multiple autocompletes, also from there, we can find input, open and close icon

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/widget-verification.spec.ts:
```diff
@@ -0,0 +1,103 @@
+    performGlobalRefresh();
+    interceptMetricsRequests().then((xhrArray) => {
+      xhrArray.forEach((xhr) => {
+        const { body: requestPayload } = xhr.request;
+        const metric = requestPayload.metric;
+        const metricData = linodeWidgets.find(
+          (data) => data.name === metric
+        );
+      if (!metricData) {
+          throw new Error(`Unknown metric: ${metric}`);
+        }
+        const granularity = requestPayload['time_granularity'];
+        const currentGranularity = granularity
+          ? granularity.value + granularity.unit
+          : '';
+        const durationUnit = requestPayload.relative_time_duration.unit.toLowerCase();
+        const durationValue = requestPayload.relative_time_duration.value;
+        const currentRelativeTimeDuration =
+          durationUnit in timeUnit
+            ? 'Last' +
+              durationValue +
+              timeUnit[durationUnit as keyof typeof timeUnit]
+            : '';
+        expect(requestPayload.aggregate_function).to.equal(
+          metricData.expectedAggregation
+        );
+        expect(currentRelativeTimeDuration).to.containIgnoreSpaces(
+          actualRelativeTimeDuration
+        );
+        expect(currentGranularity).to.containIgnoreSpaces(
+          metricData.expectedGranularity
+        );
+        expect(requestPayload.metric).to.containIgnoreSpaces(metricData.name);
```
We're on the right track with this test and I like what we're testing!

I think there's some opportunity to clean this up a bit, though:

- Let's use template literals for concatenating strings, like \`${granularity.value}${granularity.unit}\` instead of `granularity.value + granularity.unit`
    - I've noticed this done a few places throughout this PR; let's clean all of them up
- What's going on with the `timeUnit[durationUnit as keyof typeof timeUnit]`?
- The ternary on line 83 is kind of weird -- I'm having trouble understanding what condition you're trying to check for, but if it doesn't meet that condition let's throw an error and fail the test instead of assigning an empty string (kinda like what you did up on line 74)
- Why do we have to use `containIgnoreSpaces`? Shouldn't we be able to test for the exact value?

jdamore-linode commented on packages/manager/cypress/support/util/cloudpulse.ts:
```diff
@@ -0,0 +1,419 @@
+ * @param {string} service - The name of the service to select.
+ */
+export const selectAndVerifyResource  = (service: string) => {
+  const resourceInput = ui.autocomplete.findByTitleCustom('Select Resources');
```
https://docs.cypress.io/guides/references/best-practices#Assigning-Return-Values

Definitely encourage you to review this best practices doc; I've referred to it many times because it is _extremely_ helpful.

jdamore-linode commented on packages/manager/cypress/support/util/cloudpulse.ts:
```diff
@@ -0,0 +1,419 @@
+  const zoomInSelector = ui.cloudpulse.findZoomButtonByTitle('zoom-in');
+  const zoomOutSelector = ui.cloudpulse.findZoomButtonByTitle('zoom-out');
+
+  cy.get(widgetSelector).each(($widget) => {
+    cy.wrap($widget).then(($el) => {
+      const zoomInAvailable = $el.find(zoomInSelector).length > 0;
+      const zoomOutAvailable = $el.find(zoomOutSelector).length > 0;
+
+      if (zoomInAvailable) {
+        cy.wrap($el)
+          .find(zoomInSelector)
+          .should('be.visible')
+          .click({ timeout: 5000 })
+          .then(() => {
+            cy.log('Zoomed In on widget:', $el);
+          });
+      } else if (zoomOutAvailable) {
+        cy.wrap($el)
+          .find(zoomOutSelector)
+          .should('be.visible')
+          .click({ timeout: 5000 })
+          .then(() => {
+            cy.log('Zoomed Out on widget:', $el);
+          });
+      } else {
+        cy.log(
+          'Neither ZoomInMapIcon nor ZoomOutMapIcon found for widget:',
+          $el
+        );
+      }
+    });
+  });
```
Can you break this down for me a bit, since I can't see it in action?

I'm trying to understand why we're checking and reacting to whether the zoom buttons are available, instead of setting up our state in a way which would allow us to assert that they're available when they should be?

jdamore-linode commented on packages/manager/src/components/LineGraph/LineGraph.tsx:
```diff
@@ -383,6 +383,7 @@ export const LineGraph = (props: LineGraphProps) => {
               width: isLegendsFullSize ? '100%' : '85%',
             }} // this sx is added because styled table forcing the legends to be 85% width & 600px max width
             aria-label={`Controls for ${ariaLabel || 'Stats and metrics'}`}
+            data-qa-table-id={ariaLabel}
```
We have an `aria-label` on this graph, can we use `cy.findByLabelText` to select this instead of leaning on a new `data-qa-` attribute?

jdamore-linode commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -98,6 +98,7 @@ export const GlobalFilters = React.memo((props: GlobalFilterProperties) => {
             sx={{
               marginBlockEnd: 'auto',
             }}
+            data-qa-refresh-button
```
We shouldn't need this.

This button has no ARIA label, which is a bug (more specifically, it's an accessibility problem). _All_ buttons should have a label of some kind (whether visual or via `aria-label`) so that users of assistive technology can interact with it.

Once this button has been given a label, you should be able to select it using `cy.findByLabelText()`.

jdamore-linode commented on packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.tsx:
```diff
@@ -299,6 +299,9 @@ export const CloudPulseWidget = (props: CloudPulseWidgetProperties) => {
             padding={1}
           >
             <Typography
+              data-qa-widget_header={convertStringToCamelCasesWithSpaces(
```
```suggestion
              data-qa-widget-header={convertStringToCamelCasesWithSpaces(
```
Do we really need both of these?

Can we do something like `cy.get('[data-qa-widget="..."]').within(() => { /* Assert the header text exists. */ })`?

jdamore-linode reviewed (changes requested):
@agorthi-akamai Thanks for this first pass at addressing feedback!

Most of these tests are still failing in CI (it looks like we're still missing some mocks).

I've reviewed what I can so far, but once we get the tests passing I'll be able to take a closer look.

agorthi-akamai commented:
@jdamore-linode :I've addressed the comments from the code review; could you kindly evaluate.?

dwiley-akamai reviewed (commented):
`linode-widget-verification.spec.ts` passed locally ✅

agorthi-akamai commented on packages/manager/cypress/support/constants/widgets.ts:
Code review comments are adderesed

jdamore-linode commented on packages/manager/cypress/support/ui/cloudpulse.ts:
```diff
@@ -0,0 +1,26 @@
+/**
+ * cloudpulse RefreshIcon,Zoom Button UI element.
+ */
+export const cloudpulse = {
+  /**
+   * Finds a refresh button within the UI and returns the corresponding Cypress chainable object.
+   * This method is useful for locating and interacting with the refresh button in the user interface.
+   *
+   * @returns {Cypress.Chainable} - A Cypress chainable object representing the refresh button.
+   */
+  findRefreshIcon: (): Cypress.Chainable => {
+    return cy.get('[data-qa-refresh-button="true"]');
+  },
+
+/**
+ * Generates a Cypress chainable for a zoom button based on its title.
+ * This method helps in locating specific zoom buttons by their title attribute within a group of buttons.
+ *
+ * @param {string} buttonTitle - The title of the zoom button to find.
+ *
+ * @returns {Cypress.Chainable} - A Cypress chainable that represents the zoom button element.
+ */
+  findZoomButtonByTitle: (buttonTitle: string): Cypress.Chainable => {
+    return cy.get(`[data-qa-zoomer="${buttonTitle}"]`);
+  },
+};
```
```suggestion
```
I'm not willing to accept this addition because it's a workaround for a CloudPulse bug that should be fixed instead.

Buttons _have to have labels_, and if they have labels you don't need to resort to QA data attributes.

agorthi-akamai replied:
I have addressed this issue

jdamore-linode reviewed (commented):
@agorthi-akamai Hey Ananth, thanks! I see that you've applied some of the changes we discussed, I appreciate it.

However, overall I would say I'm very puzzled by the state of this PR compared to where it was when we last discussed it:

- All of the tests had been passing, but now I am once again observing consistent failures. Because I can't run the tests, I can't properly review them. I did my best to comment on what I could, but I'm blocked from further review as long as these tests are failing.
  - I've called this out before, but since we have to rely on mocks for this, can you please run your tests against an environment where CloudPulse is not enabled? It would allow you to identify mock related test failures _before_ pushing them and asking us to review, which would save yourself and us a lot of time.
- The `dashboard-filters.spec.ts` spec has been removed without any explanation
  - Overall I think it's pretty unclear what the scope of this ticket is in terms of what we're trying to cover. When we initially met and discussed Cypress tests, I suggested starting small with navigation tests. This PR has _a_ navigation test, but doesn't cover everything (for example: confirming that CloudPulse can be navigated to via the sidebar navigation, confirming that CloudPulse is absent when the feature is disabled/the account is lacking the expected capabilities, etc.). And what it _does_ cover is also somewhat unclear -- the widget verification spec has been renamed to `linode-widget-verification` so presumably only covers one type of widget. Are there tickets for other widgets? Are there tickets to restore the dashboard tests that were deleted? Are there tickets for the navigation flows I just mentioned? Tickets for error flows? The ticket description for DI-20585 is so broad that it almost sounds like it's suggesting that _all_ flows should be covered by this PR, which surely cannot be the case? This PR doesn't even approach that.

agorthi-akamai commented:
Thank you for the feedback. I'd like to provide clarification regarding the updates made in this PR:

Test Results:

I've included a screenshot for your convenience, and every test case is operating without issue.
Recent CSS modifications were the cause of the earlier failures, however those problems have now been fixed, and the tests are passing as intended.

Restructuring the test case:

The test cases that were previously included in dashboard-filters.spec.ts and widget-filters.spec.ts are contained in the linode-widget-verification.spec.ts file.
This update was done to better manage the dashboard-related test cases, especially the widget-specific ones, which are now better organised because of the reorganised tests, as we will have several dashboards.


The test coverage in this PR aligns with DI-20585, which focuses on consolidating and restructuring tests.
The  navigation and sidebar-related test cases (such as verifying CloudPulse navigation via the sidebar and checking its absence when the feature is disabled or the account lacks the expected capabilities) have now been added to this PR.


Please feel free to reach out if further clarification is needed. I appreciate your patience and ongoing feedback as we work to improve the test structure. 
![Screenshot 2024-09-25 at 8 29 42 PM](https://github.com/user-attachments/assets/89b19425-7310-49d6-97ec-a9fd76354897)

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,505 @@
+export const dashboardName = widgets.dashboardName;
+export const region = widgets.region;
+export const actualRelativeTimeDuration = timeRange.last24Hours;
+export const resource = widgets.resource;
```
I have removed the exports and unnecessary constants, and made changes based on your feedback

This version is clearer and more concise. Let me know if you need further adjustments!

agorthi-akamai commented on packages/manager/src/factories/dashboards.ts:
```diff
@@ -14,3 +66,176 @@ export const dashboardFactory = Factory.Sync.makeFactory<Dashboard>({
+      )
+    ),
+  });
+};
```
Based on your comments, I’ve updated my factory accordingly

agorthi-akamai commented on packages/manager/src/factories/widget.ts:
```diff
@@ -0,0 +1,74 @@
+import {
+  granularity,
+  timeRange,
+} from '../../cypress/support/constants/widgets';
```
I have removed the Cypress import from the app code as advised.

agorthi-akamai commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -98,6 +98,7 @@ export const GlobalFilters = React.memo((props: GlobalFilterProperties) => {
               sx={{
                 marginBlockEnd: 'auto',
               }}
+              data-qa-refresh-button
```
We have addressed this issue.

jdamore-linode replied:
Nice! That's much, much better! Thanks!

I do have one small suggestion, which would be to make the label something more human readable, e.g.

```
// Just an example suggestion, you guys might have a better idea of what the label should say.
aria-label='Refresh CloudPulse Filters'
```

This label is what gets announced to users using assistive technologies (screen readers, etc.), so we treat it like we'd treat any user facing label, rather than treating it like a purely programmatic ID.

(Your team and/or a UX designer would have a better idea of what would be a good label in this case)

But other than that small thing, this is great! Thanks again!

agorthi-akamai commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
We have addressed this issue.

jdamore-linode replied:
I'm still seeing this issue (unless it was addressed in another branch, that's cool too)

jdamore-linode replied:
Confirming that this is not a CloudPulse issue -- sorry for any time spent on this!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,505 @@
+      aclp: makeFeatureFlagData<Flags['aclp']>({ beta: true, enabled: true }),
+    }).as('getFeatureFlags');
+    mockGetAccount(mockAccount).as('getAccount'); // Enables the account to have capability for Akamai Cloud Pulse
+    mockGetFeatureFlagClientstream().as('getClientStream');
```
I have addressed this issue

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,505 @@
+  beforeEach(() => {
+    mockAppendFeatureFlags({
+      aclp: makeFeatureFlagData<Flags['aclp']>({ beta: true, enabled: true }),
+    }).as('getFeatureFlags');
```
I have addressed this issue

jdamore-linode commented on packages/manager/src/features/CloudPulse/Widget/components/Zoomer.tsx:
```diff
@@ -20,28 +20,37 @@ export const ZoomIcon = React.memo((props: ZoomIconProperties) => {
             height: '34px',
+            padding:0,
           }}
+          aria-label="zoom-in"
```
```suggestion
          aria-label="Zoom In"
```
Similar suggestion to my other comment, but something more readable for our `aria-label`s would be good. Still, this is a huge improvement. Thanks!

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,511 @@
+    // Check that the Monitor button is not present in the sidebar
+    cy.get('[data-testid="menu-item-Monitor"]').should('not.exist');
+  });
+  it('should verify the title of the widget', () => {
+    setupMethod();
+    metrics.forEach((testData) => {
+      const widgetSelector = `[data-qa-widget-header="${testData.title}"]`;
+      cy.get(widgetSelector)
+        .invoke('text')
+        .then((text) => {
+          expect(text.trim()).to.equal(
+            `${testData.title} (${testData.unit.trim()})`
+          );
+        });
+    });
+  });
```
I don't think this is a very useful test case, and I think we might still be having some trouble thinking of the tests in terms of user flows that should be covered -- instead, it feels like we've taken a list of low level technical requirements and are trying to write tests that check them off like boxes in a checklist.

A user is only going to navigate to the CloudPulse landing page because they need to accomplish some task or tasks -- what could that be? The answer to that question tells us what we're going to want to test, but I don't know the answer in this case so you guys need to ask yourselves.

(And, yes, whatever we do want to test might also require verifying that some widgets appear on the screen after the user performs some selection, and _that_ might even entail verifying that each expected widget title is visible, but that's secondary to what the test is trying to validate -- it isn't a test case in and of itself)

jdamore-linode reviewed (commented):
@agorthi-akamai Thanks for the explanations and work addressing this feedback -- this is a clear step in the right direction!

I spent most my time examining the code in `util/cloudpulse.ts` and have only looked at a few tests in detail so far: I'm sharing what I've got so far so you aren't blocked, but I'm hoping to finish looking at the rest of the tests tomorrow. 

Feel free to either wait for me to finish my review, or you're welcome to start addressing feedback.

Thanks again @agorthi-akamai!

jdamore-linode commented on packages/manager/src/features/CloudPulse/Overview/GlobalFilters.tsx:
```diff
@@ -98,6 +98,7 @@ export const GlobalFilters = React.memo((props: GlobalFilterProperties) => {
               sx={{
                 marginBlockEnd: 'auto',
               }}
+              aria-label='cloudpulse-refresh'
```
Whoops! I thought I commented on this first, but this goes along with the comments I made about the zoom in and zoom out labels: we probably want something more human readable, since this label gets announced to users of assistive technologies (screen readers, etc.).

Something like "Refresh CloudPulse" or something might make sense, but your team/the UX team would have a better idea of what makes sense here.

venkymano-akamai commented on packages/manager/cypress/support/ui/autocomplete.ts:
```diff
@@ -30,6 +30,34 @@ export const autocomplete = {
   find: (): Cypress.Chainable => {
     return cy.get('[data-qa-autocomplete] input');
   },
+  /**
+   * Finds an autocomplete input element by its placeholder text.
+   * This method is useful for locating input fields within autocomplete components
+   * when the placeholder text is known.
+   *
+   * @param {string} title - The placeholder text of the input element.
+   * @param {SelectorMatcherOptions} [options] - Optional additional options for selecting elements.
+   * @returns {Cypress.Chainable} - A Cypress chainable object that represents the located element.
+   */
+  findByPlaceholderCustom: (
+    title: string,
+    options?: SelectorMatcherOptions
+  ): Cypress.Chainable => {
+    return cy.get("[data-qa-autocomplete] input[placeholder='" + title + "']");
+  },
```
We should not need this, thanks

venkymano-akamai commented:
@jdamore-linode , Thanks for the review , The comments on autocomplete.ts findByLabel, makes sense, we will update our test cases accordingly since we have the access to the input text field now.

For the other comments, regarding aria-label and some of the best practices, we will look into it and address those.

agorthi-akamai commented:
I have addressed all code reviews, refined and restructured the code, and merged several test cases based on the feedback I received.

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,511 @@
+  ui.regionSelect.find().click().type(`${widgetDetails.linode.region}{enter}`);
+  assertSelections(widgetDetails.linode.region);
+  selectAndVerifyResource(widgetDetails.linode.resource);
+};
```
@jdamore-linode : 
Thank you for your suggestion. I have removed the setup method and opted for a step-by-step selection process to view the dashboard in each test case within the before-each method.  Hope this change the enhances readability now .
However, I’d like to highlight that before proceeding with the utility approach, I had the opportunity to present it prior to submitting the PR, but I didn’t receive any feedback at the time for this approach. Now, receiving feedback at this stage is causing us to spend additional time on rework. Moving forward, I hope we can work together to avoid this situation.

jdamore-linode replied:
> However, I’d like to highlight that before proceeding with the utility approach, I had the opportunity to present it prior to submitting the PR, but I didn’t receive any feedback at the time for this approach. Now, receiving feedback at this stage is causing us to spend additional time on rework. Moving forward, I hope we can work together to avoid this situation.

I appreciate the frustration, but this is a 1100+ line PR that's in extremely rough shape and for that reason it is going to take multiple passes of feedback and refinement in order for this to meet the standards we expect for Cloud Manager contributions.

No, I did not see this `setupMethod` last time we met over Webex and I did not notice how difficult it it was to read because at that point in time _your tests did not even pass_ and therefore I could not review them.

Once these tests are merged, they will run against every push to every PR that everybody on our team or any other team opens. They'll run every morning against our `develop` branch. They'll run multiple times every time we push a release. They'll run every time we launch a new product or datacenter. Soon they will run every time we prepare an API release. Unless you're planning to be available at all hours to assist with examining and troubleshooting CloudPulse test failures, I'm hoping you can cooperate with us now to ensure the tests are sound before we merge them.

I'd also like to point out that this PR has revealed extremely basic issues that:

- Were missed during development
- Were missed during your team's first pass at PR review
- Were missed during our team's first pass at PR review
- Were again missed during our team's comprehensive PR review
- Were observed but then _worked around_ during development of automated QA tests
- Were finally caught during PR review of the QA tests

If they hadn't been caught now, they wouldn't have been caught before launch and it would've been our customers noticing the quality issues.

Given the enormous size of these CloudPulse PRs, the amount of time it takes to review, the extent to which quality issues have _already_ been missed and have been merged, I think it is an absolutely unrealistic expectation that all CloudPulse feedback be identified and given in a single pass.

cc @jcallahan-akamai

kmuddapo replied:
@jdamore-linode  I completely understand and appreciate your points, and I want to emphasize that this isn’t about "frustration" but about ensuring we adopt the best approach for the long-term success of the project. As a team, we want to minimize back-and-forth communication and ensure our contributions meet the expected standards right from the start, as you pointed out. In today’s sync-up meeting with the Cloud Manager team, we discussed the possibility of you joining these meetings. This would give us all an opportunity to discuss approaches for complex tasks before implementing them, similar to how we currently handle feature development. Since this is the first-ever Cypress test PR for the ACLP team, it’s a great chance to align early. Let’s work together on this and ensure the tests are in great shape. Thanks again for your time.

venkymano-akamai reviewed (approved):
LGTM

venkymano-akamai commented:
> LGTM

@jdamore-linode , we have addressed most of the comments now, after that the tests and its structure looked good to me and was passing consistently. So I gave a first pass.

I know that we have cy.waits to be removed. Which is required to merge this PR.

jdamore-linode commented:
@agorthi-akamai Planning to re-review tomorrow if I can, but at a glance things look much improved. Looking forward to checking it out, I think we're almost there

agorthi-akamai commented:
@jdamore-linode: Thanks for your quick note! Happy to hear things are shaping up well. Whenever you’re able to dive into the review, I’ll be awaiting your response or any additional comments. Let me know if there’s anything further to tweak

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+  it('should trigger the global refresh button and verify the corresponding network calls', () => {
+    cy.wait(7000); //maintaining the wait since page flicker and rendering
+
+    // click the global refresh button
+    ui.button
+      .findByAttribute('aria-label', 'Refresh Dashboard Metrics')
+      .should('be.visible')
+      .click();
+
+    // validate the API calls are going with intended payload
+    cy.wait(['@getMetrics', '@getMetrics', '@getMetrics', '@getMetrics']).then(
+      (interceptions) => {
+        const interceptionsArray = Array.isArray(interceptions)
+          ? interceptions
+          : [interceptions];
+        interceptionsArray.forEach((interception) => {
+          const { body: requestPayload } = interception.request;
+          const { metric, relative_time_duration: timeRange } = requestPayload;
+          const metricData = metrics.find(({ name }) => name === metric);
+          if (!metricData) {
+            expect.fail(
+              'metricData or its expected properties are not defined.'
+            );
+          }
+          const expectedRelativeTimeDuration = timeRange
+            ? `Last ${timeRange.value} ${['hour', 'hr'].includes(timeRange.unit.toLowerCase())
+              ? 'Hours'
+              : timeRange.unit
+            }`
+            : '';
+          expect(metric).to.equal(metricData.name);
+          expect(expectedRelativeTimeDuration).to.equal(timeDurationToSelect);
+        });
+      }
+    );
```
There's a lot of room to simplify this and get rid of some conditionals.

Here's one potential solution:

```suggestion
    mockCloudPulseCreateMetrics(metricsAPIResponsePayload, serviceType).as('refreshMetrics');

    // click the global refresh button
    ui.button
      .findByAttribute('aria-label', 'Refresh Dashboard Metrics')
      .should('be.visible')
      .click();

    // validate the API calls are going with intended payload
    cy.get('@refreshMetrics.all')
      .should('have.length', 4)
      .each((xhr: unknown) => {
        const interception = xhr as Interception;
        const { body: requestPayload } = interception.request;
        const { metric, relative_time_duration: timeRange } = requestPayload;
        const metricData = metrics.find(({ name }) => name === metric);

        if (!metricData) {
          throw new Error(`Unexpected metric name '${metric}' included in the outgoing refresh API request`);
        }

        expect(metric).to.equal(metricData.name);
        expect(timeRange).to.have.property('unit', 'hr');
        expect(timeRange).to.have.property('value', 24);
      });
  });
```

A couple notes:

- I called `mockCloudPulseCreateMetrics` again and assigned it a new alias, which allows us to distinguish requests made before clicking the refresh button from those that are made after
- We currently have to cast interception objects when using `cy.get`. This is a bug in Cypress's types, and [something we've had to do elsewhere](https://github.com/linode/manager/blob/develop/packages/manager/cypress/e2e/core/notificationsAndEvents/events-fetching.spec.ts#L70-L74) in the codebase. This is an exceptional case; we usually try to avoid casting when we can (and I'm looking forward to removing these casts once the type issue is resolved)

Feel free to adapt this as needed, just wanted to demonstrate another way to go about doing this that's a little bit more readable.

agorthi-akamai replied:
I have addressed the comments accordingly. Thank you for your feedback!

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+          });
+
+          //validate the widget line graph is present and its legend rows
+          cy.findByTestId('linegraph-wrapper')
+            .should('be.visible')
+            .find('tbody tr')
+            .each(($tr) => {
+              const cells = $tr
+                .find('td')
+                .map((_i, el) => {
+                  const text = Cypress.$(el).text().trim();
+                  return text.replace(/^\s*\([^)]+\)/, '');
+                })
+                .get();
+              const [title, actualMax, actualAvg, actualLast] = cells; // the average, max and last present in the widget
+              const widgetValues = getWidgetLegendRowValuesFromResponse(
+                metricsAPIResponsePayload
+              ); // the average, max and last from the response payload
+              compareWidgetValues(
+                // compare both
+                {
+                  title,
+                  max: parseFloat(actualMax),
+                  average: parseFloat(actualAvg),
+                  last: parseFloat(actualLast),
+                },
+                widgetValues,
+                testDataTitle
+              );
+            });
+        });
```
Any chance we can clean this up a bit?

My suggestion for the `should zoom in and out of all the widgets` might not be totally applicable here, but I'm thinking that something similar should work here, too.

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+
+  it('should allow users to select desired granularity and see the most recent data from the API reflected in the graph', () => {
+    // validate the widget level granularity selection and its metrics
+    for (const testData of metrics) {
```
Any reason to use `for ... of` here instead of `metrics.forEach`?

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+      const widgetSelector = `[data-qa-widget="${testDataTitle}"]`;
+
+      cy.get(widgetSelector)
+        .first()
```
Is the `.first()` needed here? I'm thinking there should only be one of each widget on screen at a given time, right?

agorthi-akamai replied:
I have addressed the comments accordingly. Thank you for your feedback!

jdamore-linode commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    // Verifies that the expected widgets are loaded on the dashboard.
+    metrics.forEach(({ title, unit }) => {
+      const widgetSelector = `[data-qa-widget-header="${title}"]`;
+      cy.get(widgetSelector)
+        .invoke('text')
+        .then((text) => {
+          expect(text.trim()).to.equal(
+            `${title} (${unit.trim()})`
+          );
+        });
```
Can we simplify this? I think we should be able to do something like:

```typescript
cy.get(widgetSelector).should('have.text', `${title} (${unit.trim()})`).should('be.visible');
```

agorthi-akamai replied:
I have addressed the comments accordingly. Thank you for your feedback!

jdamore-linode reviewed (commented):
Thanks @agorthi-akamai! Nice work! We're getting really close.

I've left a bunch of suggestions but they're primarily concerned with improving consistency, simplification, and some miscellaneous clean up. The tests themselves are looking pretty good and seem to be running reliably, so most of my feedback there is aimed towards further increasing readability.

Not making any promises, but I think there's a good possibility that I'll be ready to approve this PR once these suggestions have been addressed. I also see that the flickering PR is approved, so perhaps we'll be able to get that merged and see if we can get rid of the `cy.wait`s, too.

Thanks again @agorthi-akamai, this is shaping up nicely!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/cloudpulse-navigation.spec.ts:
You're welcome! I'm glad to hear that the spec looks and runs great!

agorthi-akamai commented on packages/manager/cypress/support/constants/widgets.ts:
```diff
@@ -0,0 +1,100 @@
+ * Defines the granularity levels used for specifying time intervals in data aggregation or reporting.
+ * Each property represents a different granularity level.
+ */
+
```
/Removed doc comment for granularity and added a descriptive comment for widgetDetails

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,134 @@
+    apiMatcher(`/monitor/services/${service_type}/token`),
+    JWSToken
+  );
+};
```
I have addressed the issue by adding const mockToken = token ?? randomString(62);, which allows the token to be an optional parameter. Thank you for the suggestion!

agorthi-akamai commented on packages/manager/src/factories/dashboards.ts:
```diff
@@ -14,3 +53,175 @@ export const dashboardFactory = Factory.Sync.makeFactory<Dashboard>({
+  }
+);
+// Function to generate random values based on the number of points
+export const generateValues = (
```
I have renamed the function to generateRandomMetricsData to better reflect the type of data it generates. Additionally, I have moved it to cypress/support/util/cloudpulse.ts,. Thank you for your feedback!

agorthi-akamai commented on packages/manager/src/factories/dashboards.ts:
```diff
@@ -14,3 +53,175 @@ export const dashboardFactory = Factory.Sync.makeFactory<Dashboard>({
+    intervals[granularityData] ||
+    (() => {
+      throw new Error(`Unsupported granularity: ${granularityData}`);
+    })();
```
I have addressed your suggestion by creating a string literal type for granularityData with the values '1 day' | '1 hr' | '5 min' | 'Auto'. This change allows TypeScript to handle type checks, streamlining the code and eliminating the need for manual validation. Thank you for the great suggestion!

agorthi-akamai commented on packages/manager/src/factories/dashboards.ts:
```diff
@@ -14,3 +53,175 @@ export const dashboardFactory = Factory.Sync.makeFactory<Dashboard>({
+    timeRanges[time] ||
+    (() => {
+      throw new Error(`Unsupported time range: ${time}`);
+    })();
```
I have addressed the comments accordingly. Thank you for your feedback!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    average: expectedAverage, last: expectedLast,
+  } = expectedValues;
+
+
```
I have addressed the comments accordingly. Thank you for your feedback!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+
+          // validate the widget details
+          cy.findByTestId('linegraph-wrapper')
+            .as('canvas')
```
I have addressed the comments accordingly. Thank you for your feedback!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,498 @@
+  const extractedTitle = actualTitle.substring(
+    0,
+    actualTitle.indexOf(' ', actualValues.title.indexOf(' ') + 1)
+  );
```
I have addressed the feedback by implementing the change where actualTitle now starts with the expected title

agorthi-akamai commented:
Hi Joe!
I’ve addressed your code review feedback and added appropriate comments in the pull request. Additionally, I’ve removed the waits as per our preference since the PR has been merged. Could you please take a moment to review it?

jdamore-linode reviewed (commented):
Thanks Ananth! A few things I suggested before that still need to be addressed:

- I think there may have been a [misunderstanding](https://github.com/linode/manager/pull/10891#discussion_r1781772446): we should be preferring `forEach` over `for .. of` expressions, so we'll want to change all of these new `for .. of` expressions back to `forEach`. I apologize for any confusion.
- These two tests still need some clean up:
  - `should allow users to select the desired aggregation and view the latest data from the API displayed in the graph`
  - `should allow users to select their desired granularity and see the most recent data from the API reflected in the graph`
- Order of params in the mock utils in `intercepts/cloudpulse.ts` (I posted a new comment that goes into more detail about this, but feel free to reach out on Slack if you have any questions or if you're not sure what I'm asking for. It's a quick change.)

Beyond that, I noticed a couple small things that I failed to catch before (sorry!) and shared some suggestions to address them:

- Stray `cy.get()` after `cy.visitWithLogin()` -- double check that this isn't required, but I'm pretty sure it's unnecessary. It doesn't hurt anything, though, so not a huge deal.
- Couple naming suggestions for the functions in `util/cloudpulse.ts` to make them consistent with the rest of our mock and intercept utils

There's good news though:

- The tests pass reliably: I ran them 25 times in a row, no failures. I'm very confident in their stability right now, and am very happy to see them running so smoothly (and quickly) without the need for `cy.wait`
- Almost all of these remaining suggestions are about maintaining consistency with our codebase's conventions -- only the clean up for the 2 tests I mentioned above actually relate to test quality. In other words, we're pretty much there in terms of test quality.
- Once these suggestions are addressed, as long as there are no regressions and the tests continue passing reliably, I'm ready to approve.

Thanks Ananth! We're just about there. Don't hesitate to reach out if you have any questions or could use a hand wrapping up these last few things.

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
+ * @returns {Cypress.Chainable<null>} - Returns a Cypress chainable object, enabling command chaining in tests.
+ */
+
+export const mockCloudPulseJWEToken = (
```
I have addressed the comments accordingly. Thank you for your feedback!

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
+ * @returns {Cypress.Chainable<null>} - Returns a Cypress chainable object, allowing for command chaining in tests.
+ */
+
+export const mockCloudPulseDashboardServicesResponse = (
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
+ * @param {any} mockResponse - The mock response to return for the intercepted request.
+ * @returns {Cypress.Chainable<null>} The chainable Cypress object.
+ */
+export const mockCloudPulseCreateMetrics = (
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
++ * @returns The chainable Cypress object.
++ */
+
+export const mockCloudPulseGetDashboards = (
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
+ * @returns {Cypress.Chainable<null>} The chainable Cypress object.
+ */
+
+export const mockCloudPulseGetMetricDefinitions = (
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,401 @@
+
+    // navigate to the cloudpulse page
+    cy.visitWithLogin('monitor/cloudpulse').as('cloudPulsePage');
+    cy.get('@cloudPulsePage');
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,401 @@
+            cy.findByText( `${expectedWidgetValues.max} ${testData.unit}`).should('be.visible');
+            cy.findByText(`${expectedWidgetValues.average} ${testData.unit}` ).should('be.visible');
+            cy.findByText( `${expectedWidgetValues.last} ${testData.unit}`).should('be.visible');
+          });
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,401 @@
+    }
+  });
+  it('should allow users to select the desired aggregation and view the latest data from the API displayed in the graph', () => {
+    for (const testData of metrics) {
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -0,0 +1,141 @@
+
+export const mockCloudPulseGetMetricDefinitions = (
+  metricDefinitions: MetricDefinitions,
+  serviceType: string
```
I have addressed the comments accordingly

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,401 @@
+              ); // the average, max and last from the response payload
+              compareWidgetValues( { title,max: parseFloat(actualMax),average: parseFloat(actualAvg),last: parseFloat(actualLast)},
+              expectedWidgetValues, testData.title );
+            });
```
I have followed the zoom in/zoom out testing approach

jdamore-linode commented on packages/manager/cypress/support/intercepts/cloudpulse.ts:
```diff
@@ -16,7 +16,7 @@ import type {
 } from '@linode/api-v4';
 
 /**
- * Intercepts  GET requests for metric definitions.
```
Nice catch! 🦅👀

jdamore-linode reviewed (approved):
Nice work, thanks for the effort you put into this @agorthi-akamai.

I posted a couple small suggestions that I'd like to see applied, and think there might be a stray inclusion in `queries/cloudpulse/metrics.ts` that might not belong in this PR, but otherwise I'm quite happy with the state of this. Thanks again

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,376 @@
+      .click();
+
+    cy.findByText(resource).should('be.visible');
+    cy.wait(['@getMetrics', '@getMetrics', '@getMetrics', '@getMetrics']);// Wait for all metrics query requests to resolve. 
```
I have addressed the comments accordingly. Thank you for your feedback!

agorthi-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/linode-widget-verification.spec.ts:
```diff
@@ -0,0 +1,376 @@
+    // navigate to the cloudpulse page
+    cy.visitWithLogin('monitor/cloudpulse');
+    cy.wait('@fetchServices'); // Wait for services
+    cy.wait('@fetchDashboard'); // Wait for dashboard
```
I have addressed the comments accordingly. Thank you for your feedback!

jdamore-linode replied:
Looks like this one might've slipped through the cracks (as well as in the DBaaS spec)

jdamore-linode replied:
I'm sorry about that! I was looking at an old revision, we're good to go here. Thanks Ananth!

agorthi-akamai commented on packages/manager/src/queries/cloudpulse/metrics.ts:
```diff
@@ -38,7 +38,7 @@ export const useCloudPulseMetricsQuery = (
 
     enabled: !!obj.isFlags,
-    refetchInterval: 120000,
+    refetchInterval: 1800000,
```
My apologies, that was included by mistake. I've already reverted the changes.

jdamore-linode replied:
Thanks!

dwiley-akamai reviewed (commented):
Things look good to me, but there's a unit test failure for `GlobalFilters.test.tsx` -- should be able to resolve it by updating

https://github.com/linode/manager/blob/92a4643692f5b0f1c5b328c86cf579193ab322a1/packages/manager/src/features/CloudPulse/Overview/GlobalFilters.test.tsx#L38

to have `name: 'Select a Time Duration'`

agorthi-akamai commented:
All comments have been addressed.

agorthi-akamai commented:
All comments are addressed.

dwiley-akamai reviewed (approved):
All checks pass now ✅

Thanks for all the changes and fixes!
